docs/style-guide.md

```diff
@@ -0,0 +1,423 @@
+
+Note however that none of this document is meant to trump common sense, if if you're in a situation where it would be better to violate a **required** practice, then it just means we should have a discussion about it. There will almost certainly be many cases where we need to update this doc and even go back to existing code to update it in the light of new information. However exceptions should always be commented clearly for the next maintainer.
+
+## Sources
+
+* [Learn Rust](https://doc.rust-lang.org/)
+  * The canonical source for Rust information, but it doesn't offer advice on all aspects of code design.
+* [Introduction - Learning Rust With Entirely Too Many Linked Lists](https://rust-unofficial.github.io/too-many-lists/)
+  * This is brilliantly written and very educational about how you get into a Rust mindset.
+* [Elegant Library APIs in Rust](https://deterministic.space/elegant-apis-in-rust.html)
+  * This has a lot of good points and is well worth a read, but be warned that some of the details about implementation are somehwat out of date (2017). It has a video too.
+* [Good Practices for Writing Rust Libraries](https://pascalhertleif.de/artikel/good-practices-for-writing-rust-libraries/)
+  * Shorter and more focussed on the act of coding the libraries, and less about API design. Also potentially out of date in places.
+* [Strategies for Returning References in Rust](http://bryce.fisher-fleig.org/blog/strategies-for-returning-references-in-rust/index.html)
+  * Though I don't believe we should be doing this, it's still an interesting read.
+
+## Other Useful Links
+
+* Write and run Rust snippets: https://play.rust-lang.org
+  * You can save snippets in permantent links and incluce them as working examples in docs.
+* Write code and see what it compiles to: https://rust.godbolt.org
+  * Note that you need to enable compiler optimizations via `-C opt-level=3` if you are looking to meaningfully compare two code snippets.
+
+# Naming Conventions
+
+### No Special Naming :: required
+
+Follow the naming advice in [Naming - Rust API Guidelines](https://rust-lang.github.io/api-guidelines/naming.html) at all times, and use [Rust RFC 0430](https://github.com/rust-lang/rfcs/blob/master/text/0430-finalizing-naming-conventions.md) for any ambiguous issues (e.g. Crate naming).
+
+Based on [this comment](https://github.com/rust-lang/api-guidelines/issues/29#issuecomment-342745898) I think the advice for Crate names should be:
+
+* Infer Crate names from package names where feasible:
+  * A package is the directory in which the Cargo.toml exists.
+* Prefer single (lowercase) words for package names where feasible
+* Use __lowercase with underscore__ (a.k.a "snake_case") to split words in package names if needed
```

Comment:
overall my impression is that it's more normal to use dashes in the crate name, and those get automatically converted to underscores in rust code

underscores in crate names are more prevalent in very old crates from before this era

Reply:
So, I thought that there was a distinction between "Crate names" and "Package names". And the one had to be a valid Rust identifier, which would seem to preclude '-' in one of these cases.

Am I mistaken?

I'm happy with whichever works, and if we have "Package name = foo-bar" ==> Crate name: foo_bar", then that's still 100% unamibguous so that's fine.

Reply:
There's a slight distinction but it's unimportant here. Dashes in crate names automatically turn into underscores in Rust.

Reply:
(I was wrong. Either way, we can use kebab-case in package names and the crate name will automatically have underscores without extra work. Rustaceans are used to this and will expect it)